Refactor service trackers

lpapazow opened the pull request:
As discussed [here](https://github.com/openhab/static-code-analysis/issues/265), I added these abstract classes as suppressions for the `DeclarativeServicesDependencyInjectionCheck`. 

This check was also triggered by the `RESTActivator` class. Instead of having a `localeProviderTracker` in the `RESTActivator` and have a `LocaleUtil` class that uses the RESTActivator's `getLocale`, I changed `LocaleUtil` to use DS and removed the `RESTActivator`.

If you think this change is unnecessary or potentially dangerous, I could revert it and just add the RESTActivator class as a suppression.

Signed-off-by: Lyubomir V. Papazov <lpapazow@gmail.com>

maggu2810 commented:
As you already refactored it, why not doing it a little bit more the OSGi way.
* drop immediate activation of `LocaleService`
* make the `Locale getLocale(String acceptLanguageHttpHeader)` method non static
* the components / REST resources could require a mandatory reference for `LocaleService` (as the locale service itself is using a dynamic reference to the provider it can be activated all the time)

No need for an activator or access a static method of a service class.

WDYT?

lpapazow commented:
@maggu2810, I agree completely, having a mandatory reference and non staitic methods in the LocaleService seems like a way better idea.

@htreu, after this [commit](https://github.com/eclipse/smarthome/commit/b3933fa4ca29e6624db8f8b98f35552afbf3a3bb) the `getContext` method is no longer used and the only place the `getLocale` method is used is the LocaleUtil class. Could you please clarify if this is what you mean by API breaking and do you think refactoring the code in the way maggu proposed would be ok?

htreu commented:
@lpapazow yes you are right. I just realised that RESTActivator is internal. And I second @maggu2810's idea to drop static stiff completely and make use of the new component throughout the REST components.

lpapazow commented:
I added a mandatory reference for LocaleService in the REST resources. Some of them didn't use the `@Component` annotation but used the xml files in `OSGI-INF` folder instead. I refactored all of the classes that now require LocaleService to use annotations , and then I refactored all the other classes in those bundles to use annotations, deleted all the redundant .xml files and added .gitignore files in a seperate commit.

maggu2810 commented on bundles/automation/org.eclipse.smarthome.automation.rest/META-INF/MANIFEST.MF:
```diff
@@ -17,5 +17,6 @@ Import-Package:
  org.eclipse.smarthome.config.core,
  org.eclipse.smarthome.core.common.registry,
  org.eclipse.smarthome.io.rest,
+ org.osgi.service.component.annotations,
```
IIRC there has been a commit some time ago that removes all that imports.
Annotations does not need to be present at runtime (by the JVM) and the one here are known to be used at build time only.

maggu2810 commented on bundles/automation/org.eclipse.smarthome.automation.rest/src/main/java/org/eclipse/smarthome/automation/rest/internal/ModuleTypeResource.java:
```diff
@@ -73,6 +80,15 @@ protected void unsetModuleTypeRegistry(ModuleTypeRegistry moduleTypeRegistry) {
         this.moduleTypeRegistry = null;
     }
 
+    @Reference(cardinality=ReferenceCardinality.MANDATORY, policy=ReferencePolicy.STATIC)
```
`cardinality=ReferenceCardinality.MANDATORY, policy=ReferencePolicy.STATIC` is default, so no need to set it explicit.
Can you remove it here and on other places

htreu commented on bundles/io/org.eclipse.smarthome.io.rest.core/src/main/java/org/eclipse/smarthome/io/rest/core/internal/item/ItemResource.java:
```diff
@@ -176,6 +178,15 @@ protected void unsetDTOMapper(DTOMapper dtoMapper) {
+    }
+    
+    protected void unsetLocaleService(LocaleService localeService) {
+        this.localeService = null;
```
with this you should have an error in your IDE: `Null type mismatch: required '@NonNull LocaleService' but the provided value is null`

htreu commented on bundles/io/org.eclipse.smarthome.io.rest.core/src/main/java/org/eclipse/smarthome/io/rest/core/internal/item/ItemResource.java:
```diff
@@ -130,6 +131,7 @@
     @NonNullByDefault({})
     private DTOMapper dtoMapper;
     private final Set<ItemFactory> itemFactories = new HashSet<>();
+    private LocaleService localeService;
```
Here you should have an error in your IDE: `The @NonNull field localeService may not have been initialized`
You can circumvent this by adding the `@NonNullByDefault({})` annotation here. This will basically exclude the field from the null annotation checks.

htreu commented on bundles/io/org.eclipse.smarthome.io.rest.core/src/main/java/org/eclipse/smarthome/io/rest/core/internal/item/ItemResource.java:
```diff
@@ -65,7 +65,8 @@
 import org.eclipse.smarthome.io.rest.DTOMapper;
 import org.eclipse.smarthome.io.rest.JSONResponse;
-import org.eclipse.smarthome.io.rest.LocaleUtil;
+import org.eclipse.smarthome.io.rest.LocaleServiceImpl;
```
this import is reported to be unsed, please remove.

htreu commented on bundles/io/org.eclipse.smarthome.io.rest/src/main/java/org/eclipse/smarthome/io/rest/LocaleService.java:
```diff
@@ -0,0 +1,9 @@
+package org.eclipse.smarthome.io.rest;
```
license header missing

htreu commented on bundles/io/org.eclipse.smarthome.io.rest/src/main/java/org/eclipse/smarthome/io/rest/LocaleService.java:
```diff
@@ -0,0 +1,9 @@
+
+import java.util.Locale;
+
+public interface LocaleService {
```
JavaDoc & author tag missing

htreu commented on bundles/io/org.eclipse.smarthome.io.rest/src/main/java/org/eclipse/smarthome/io/rest/LocaleService.java:
```diff
@@ -0,0 +1,9 @@
+
+public interface LocaleService {
+
+    public Locale getLocale(String acceptLanguageHttpHeader);
```
methods on public interfaces are public by default, please remove the keyword.

htreu commented:
Lets wait for rebase and Travis, I guess we are good to go then.

sjka commented:
@lpapazow friendly ping

lpapazow commented:
Thanks for the reminder, I've rebased the branch.

maggu2810 commented:
@htreu Have your requested changes been applied.

maggu2810 commented:
> All injected services to a REST resource have to be referenced optional/dynamic to allow the resource to resolve anytime.

Why?

sjka commented:
To avoid having jersey rebuilds the entire universe again and again - see #2320

htreu commented:
>> All injected services to a REST resource have to be referenced optional/dynamic to allow the resource to resolve anytime.

> Why?

The REST framework jersey is not able to restart single resources in case a dependency gets registered/unregistered. So all resources get restarted every time dependent services register/unregister. The relevant change was introduced with https://github.com/eclipse/smarthome/pull/2320.

maggu2810 commented:
Thank you both for point me the respective PR.
Now I remember again. :wink:

lpapazow commented:
I made the requested changes, sorry for taking so long.

maggu2810 commented on bundles/io/org.eclipse.smarthome.io.rest/src/main/java/org/eclipse/smarthome/io/rest/LocaleServiceImpl.java:
```diff
@@ -0,0 +1,93 @@
+
+    private static final Logger logger = LoggerFactory.getLogger(LocaleServiceImpl.class);
+
+    private static LocaleProvider localeProvider;
```
* remove `static`
* add `volatile`

htreu replied:
Do you claim `volatile` here because `localeProvider` is a dynamic reference and can be changed any time? Other threads holding a LocaleService need to check with the main copy then, correct?

maggu2810 replied:
Correct.
For dynamic references volatile or atomic references should be used.
If you want to spend your time in reading some discussion about: https://github.com/eclipse/smarthome/pull/3617
:wink:

maggu2810 commented on bundles/io/org.eclipse.smarthome.io.rest/src/main/java/org/eclipse/smarthome/io/rest/LocaleServiceImpl.java:
```diff
@@ -0,0 +1,93 @@
+@Component
+public class LocaleServiceImpl implements LocaleService {
+
+    private static final Logger logger = LoggerFactory.getLogger(LocaleServiceImpl.class);
```
* remove `static`

maggu2810 commented on bundles/automation/org.eclipse.smarthome.automation.rest/src/main/java/org/eclipse/smarthome/automation/rest/internal/TemplateResource.java:
```diff
@@ -48,13 +52,16 @@
  */
 @Path("templates")
 @Api("templates")
+@Component(service = { RESTResource.class, TemplateResource.class })
```
Why do you need to export the the implementation class itself as service (has not been done before)?

lpapazow replied:
I changed it because of this [comment](https://github.com/eclipse/smarthome/pull/5265#discussion_r188225858)

htreu replied:
Sorry for the confusion. @maggu2810 is completely right, the REST services should be declared to provide only the RESTResource interface. Please update accordingly @lpapazow. Thanks.

kaikreuzer replied:
Sorry to comment so late, but I would actually prefer if you could remove all changes to the automation engine from this PR, because it will cause huge conflicts for #4468, which is already addressing these components as well.

lpapazow replied:
If I want to keep the LocaleService logic, I should injected the locale into the automation RESTResources. Would it be okay to add refererences in the .xml files in the OSGI-INF folder?

kaikreuzer replied:
Oh, I didn't look closely enough to notice that it all is linked to the LocaleService...
No, changing the XML won't be any better wrt conflicts. I guess it is then ok to leave this PR as it is and rather rebase #4468 later on and sync the changes again. @adimova do you agree?

maggu2810 commented on bundles/automation/org.eclipse.smarthome.automation.rest/src/main/java/org/eclipse/smarthome/automation/rest/internal/ModuleTypeResource.java:
```diff
@@ -58,13 +62,16 @@
  */
 @Path("module-types")
 @Api("module-types")
+@Component(service = { RESTResource.class, ModuleTypeResource.class })
```
Why do you declare and so export the class itself as provided service?

lpapazow replied:
Changed because of [this](https://github.com/eclipse/smarthome/pull/5265#discussion_r188215435)

maggu2810 commented on bundles/automation/org.eclipse.smarthome.automation.rest/src/main/java/org/eclipse/smarthome/automation/rest/internal/RuleResource.java:
```diff
@@ -76,6 +80,7 @@
  */
 @Path("rules")
 @Api("rules")
+@Component(service = { RESTResource.class, RuleResource.class })
```
Why do you declare and so export the class itself as provided service?

maggu2810 commented on bundles/io/org.eclipse.smarthome.io.rest.voice/src/main/java/org/eclipse/smarthome/io/rest/voice/internal/VoiceResource.java:
```diff
@@ -55,7 +55,7 @@
  * @author Kai Kreuzer - Initial contribution and API
  */
-@Component
+@Component(service = { RESTResource.class, VoiceResource.class })
```
Why do you declare and so export the class itself as provided service?

kaikreuzer commented:
@lpapazow Any update on this?

htreu commented on bundles/io/org.eclipse.smarthome.io.rest.sitemap/META-INF/MANIFEST.MF:
```diff
@@ -32,6 +32,8 @@ Import-Package:
  org.eclipse.smarthome.ui.items,
  org.glassfish.jersey.media.sse,
  org.glassfish.jersey.server,
+ org.osgi.service.component.annotations,
```
This dependency is not needed, please remove.

htreu commented on bundles/io/org.eclipse.smarthome.io.rest.sitemap/src/main/java/org/eclipse/smarthome/io/rest/sitemap/internal/SitemapResource.java:
```diff
@@ -108,7 +108,7 @@
  * @author Yordan Zhelev - Added Swagger annotations
  */
-@Component(service = { SitemapResource.class, RESTResource.class })
+@Component
```
This will expose the service with all attached interfaces. Please use `@Component(service = RESTResource.class )` here.

htreu commented on bundles/ui/org.eclipse.smarthome.ui.icon/src/main/java/org/eclipse/smarthome/ui/icon/internal/IconServlet.java:
```diff
@@ -41,6 +45,7 @@
  *
  * @author Kai Kreuzer - Initial contribution
  */
+@Component
```
The original service xml stated also `configuration-policy="optional"` and `modified="modified"`.

htreu commented on bundles/model/org.eclipse.smarthome.model.script/OSGI-INF/voiceaction.xml:
```diff
@@ -1,22 +0,0 @@
-<?xml version="1.0" encoding="UTF-8"?>
```
This file is removed, but I'm missing the annotation at `VoiceActionService`.

htreu commented:
Only Travis complains. Unfortunately I don't understand why.

kaikreuzer commented:
I have restarted Travis, now it is green.
@maggu2810 As you still have a "request changes" review open, I leave merging to you.

maggu2810 commented on bundles/ui/org.eclipse.smarthome.ui.icon/src/main/java/org/eclipse/smarthome/ui/icon/internal/IconServlet.java:
```diff
@@ -41,6 +47,7 @@
  *
  * @author Kai Kreuzer - Initial contribution
  */
+@Component(configurationPolicy = ConfigurationPolicy.OPTIONAL)
```
> If not specified, the OPTIONAL configuration policy is used. 

On no other places we set the configuration policy to optional explicit, why here?

htreu replied:
It was defined like that in the original service XML, see https://github.com/eclipse/smarthome/pull/5265#discussion_r197353799.

In case its optional I suggest addressing this in another PR for better traceability.

maggu2810 replied:
@htreu What do you want to trace?
We migrate to the component annotations.
The annotations have default values, that does not need to be set explicitly.
We have requested to drop implicit values all the time (in this PR and all the others) if we found their usage in the review. Don't you remember on the comments about dropping the mandatory (1..1) reference and the static policy because it is the default value of the annotation? It does not matter, if this has been part of the XML before (e.g. https://github.com/eclipse/smarthome/pull/5265/files#diff-46cb4a3fb85d0b0057a3f9264fc8da74L18).

htreu replied:
Alright, thanks for the implicit explanation that `configurationPolicy = ConfigurationPolicy.OPTIONAL` is the default :-)
@lpapazow please drop it as requested.

maggu2810 commented on bundles/io/org.eclipse.smarthome.io.rest.sitemap/META-INF/MANIFEST.MF:
```diff
@@ -35,3 +35,4 @@ Import-Package:
  org.osgi.framework,
  org.slf4j
 Service-Component: OSGI-INF/*.xml
+Bundle-ActivationPolicy: lazy
```
Not needed

htreu commented:
Thanks @lpapazow for the update.
@maggu2810 this one should be good to go.

sjka commented:
I flagged this as APIBreaking because of the vanished `LocaleUtil`.
